Gradual reservation of minor heap adresses

gasche opened the pull request:
This PR proposes a fix for #14153 . Instead of reserving minor-heap address space for 128 domains at once, it start by reserving enough address space for 1 domain, then 8, 16, 32, 64, etc. This is useful for programs that want to use large minor heaps, and run in environments that refuse large address space reservations. (Apparently Rocq is hit by this limitation in some settings.)

The PR sits on top of both #14160 (a refactoring changing naming conventions about the minor heap reservation, suggested by @kayceesrk) and #14161 (a refactoring that makes it possible to reserve domain structures before initialization, without adding them as STW participants yet). I am marking it as a "draft" with the intention to encourage reviewing and merging those PRs first, so that the diff here becomes simpler and thus the PR easier to review.

In the present PR:
- we change the `Domain.spawn` process so that the *parent* domains reserves the domain structure for the child, by making it a 'parked' domain in the terminology of #14161 (this is fairly simple)
- then we add a mechanism to allocate the minor heaps reservation gradually; when the parent reserves a domain structure, it may need to increase the size the reservation.

The growth strategy implemented in the PR is to first reserve virtual address space only for 1 domain (the common case for most OCaml 5 programs), then 8, then 16, 32, 64... doubling until `caml_params->max_domains` is reached. (I could do 1, 2, 4, 8, but my guesstimate was that most programs that spawn at least one domain actually use as many as CPUs, which is typically in the 4-8 range on today's machines.)

The change is relatively simple, but note that the mechanism for updating the minor heaps reservation was very rarely used before (it would only kick in if the user increases the size of the minor heap while running), and now it runs on every program that spawns domains, so we may hit issues with it that we had not noticed before. (I already found and fixed something odd, #14158, while working on this.)

gasche committed:
domain_create: have the new domain reserved by the parent

This will help to implement on-demand resizing of the minor heaps
reservation, as the parent can pause 'domain_spawn' to do a new
reservation (whereas the child cannot run a STW section before it is
fully initialized).

gasche committed:
domain: gradually reserve the minor-heaps memory space

kayceesrk commented:
Thanks @gasche. I’ll wait until this is no longer a draft before I have a look.

silene commented on runtime/domain.c:
```diff
@@ -242,77 +242,149 @@ CAMLexport atomic_uintnat caml_num_domains_running = 0;
+  Three regions in the [domains]  array correspond to three states for domains:
+   - at indices [0, active_domains) are the 'active' domains,
+     which participate in STW sections.
+   - [particiating_domains, parked_domains) are the 'parked' domains,
```
```suggestion
   - [active_domains, parked_domains) are the 'parked' domains,
```

gasche replied:
Thanks! Fixed in #14161.

gasche committed:
Changes

gasche commented:
I have rebased this PR now that #14160 and #14161 have been reviewed and merged by @kayceesrk (thanks!). It is implemented as two independent commits that can be reviewed independently.

I see test failures in testsuite/tests/parallel/domain_parallel_spawn_burn_gc_set.ml, which was added precisely to exercise minor heap reservation changes. I still need to study the failures and push a fix.

gasche committed:
check_minor_heap: avoid noise on non-debug builds, strengthen checks

gasche commented:
I have fixed the failure in testsuite/tests/parallel/domain_parallel_spawn_burn_gc_set.ml and rebased the PR. (When debugging this issue I encountered an independent issue with a two-line proposed fix in #14164, and the present PR is currently rebased on top of that PR.)

gasche commented:
@kayceesrk merged the fix in #14164 (thanks!), I rebased the current PR on top of trunk and it is ready for review again, there are no known remaining issues.

kayceesrk commented:
https://github.com/ocaml/ocaml/actions/runs/16490544346/job/46624400162?pr=14162#step:6:815

This looks like something to be fixed?

gasche commented:
Thanks, I'll have a look.

gasche committed:
domain.c: the minor heap reservation size only grows

gasche committed:
domain.c: get up-to-date caml_minor_heap_max_wsz value

gasche committed:
runtime: useless write in gc_ctrl.c

The value of [caml_minor_heap_max_wsz] is already correctly set by the
call to [reserve_minor_heaps_reservation] from [caml_init_domains]. It
is simpler to reason about code if the logic to modify this part of
theruntime state is concentrated in domain.c.

gasche commented:
I cannot reproduce the failure on my machine, so it is hard to be certain, but here is my guess as to what went wrong, and the changes I made accordingly:

- two domains concurrently tried to change the major heap reservation, one (A) to increase the number of reserved domains and the other (B) to increase the reservation size
- domain (B) went first and increased the reservation to a new, larger size
- domain (A) went second and set back the reservation size to its initial, smaller value
  (this comes from the fact that the value of `caml_minor_heap_max_wsz` was read before the STW-loop, instead of an up-to-date value being read at each iteration; I fixed this)

The current runtime code was not completely clear on whether it is okay to shrink the minor heap reservation size (it would accept to shrink it), but I believe that it was not correct in case it shrinks at runtime -- we would need to also shrink the minor heap arenas accordingly, and we do not do this. I changed the code to guarantee that the reservation size only grows, smaller sizes get ignored. (This would, in itself, also fix the debug-runtime assertion failure we observed.)

I updated the PR with three new commits that try to fix this issue.

gasche commented:
There is another testsuite failure with the debug runtime, in lib-dynlink-domains/main.ml:

```
[03] file runtime/domain.c; line 287 ### Assertion failed: has_interrupt_word
```

It looks like I need to learn how to run the testsuite with the debug runtime enabled :-)

gasche committed:
testsuite/HACKING.adoc: document USE_RUNTIME

Octachron commented:
`USE_RUNTIME=d make parallel` will select the debug runtime when running the testsuite.

gadmm commented:
There seems to be conflicts with #12410 which has been ready for review since July 3rd, and was rewritten several times to accommodate concurrent changes in the runtime. I was wondering whether you might have a plan for rebasing #12410 on top of this one, or if it might be better to first review and merge #12410 before the current PR.

gasche commented:
The last testsuite failure is interesting, related to the handling of domains around STW registration. I will make the present message self-contained because the issue is different from what was discussed so far in the thread, and touches code that was written by @gadmm who probably did not follow earlier discussions.

In trunk before #14161, we had a piece of code that runs in debug mode when we add a new domain as a STW participant:

```c
static void add_next_to_stw_domains(void)
{
  CAMLassert(stw_domains.participating_domains < caml_params->max_domains);
  stw_domains.participating_domains++;
#ifdef DEBUG
  /* Enforce here the invariant for early-exit in
     [caml_interrupt_all_signal_safe], because the latter must be
     async-signal-safe and one cannot CAMLassert inside it. */
  bool prev_has_interrupt_word = true;
  for (int i = 0; i < caml_params->max_domains; i++) {
    bool has_interrupt_word = all_domains[i].interruptor.interrupt_word != NULL;
    if (i < stw_domains.participating_domains) CAMLassert(has_interrupt_word);
    if (!prev_has_interrupt_word) CAMLassert(!has_interrupt_word);
    prev_has_interrupt_word = has_interrupt_word;
  }
#endif
}
```

(The invariant remains checked after #14161, in a `check_stw_domains` function that is called more often, but this is irrelevant to the discussion below.)

The comment about the check refers to the function that runs to interrupt all domains, and can run in arbitrary contexts including from a signal handler:

```c
/*  This function is async-signal-safe as [all_domains] and
    [caml_params->max_domains] are set before signal handlers are installed and
    do not change afterwards. */
void caml_interrupt_all_signal_safe(void)
{
  for (dom_internal *d = all_domains;
       d < &all_domains[caml_params->max_domains];
       d++) {
    /* [all_domains] is an array of values. So we can access
       [interrupt_word] directly without synchronisation other than
       with other people who access the same [interrupt_word].*/
    atomic_uintnat * interrupt_word =
      atomic_load_acquire(&d->interruptor.interrupt_word);
    /* Early exit: if the current domain was never initialized, then
       neither have been any of the remaining ones. */
    if (interrupt_word == NULL) return;
    interrupt_domain(&d->interruptor);
  }
}
```

That interruption function relies on the fact that it can terminate when encountering an `interrupt_word` that is `NULL`, the assumption being that all domains that follow it in `all_domains` order will also have a NULL interrupt word -- which means that they have never been activated to participate to STW events.

The "invariant" checked in debug mode asks that:

1. Active/participating domains have a non-NULL interrupt word.
2. NULL-interrupt-word domains form a suffix of the `active_domains`, there is no non-NULL domain after a NULL domain.

Interrupt words are initialized to `NULL` when we initialize the runtime, and `domain_create` makes it non-null during the `all_domains_lock` critical section in which the domain is added as a STW participant. They are never set back to NULL while the runtime is running. This guarantees part (1) of the invariant above: we add the domain as a STW participant after making its interrupt word non-NULL. The part (2) of the invariant holds for a more subtle reason: when spawning a domain, we always pick the first non-active domain, which is sometimes a domain that has already been initialized before (so its interrupt word is already non-NULL) or a domain that has not been initialized before, but in that case we are in the suffix of `stw_domains` that is in the same order as `all_domains`, and so the first never-activated domain in `stw_domains` order is also the first never-activated domain in `all_domains` order.

In this PR, I changed the domain-structure reservation logic so that the caller of `Domain.spawn` (the parent domain) reserves the domain structure that will be used for their child. In particular, it is possible for two domains to call `Domain.spawn` in parallel, and to reserve (park, in the terminology of #14161) two domain structures in parallel. One of them will win `all_domains_lock` and initialize its child domain first in the reserved structure. If the *second* child domain in `all_domains` order wins the lock, then that domain will be initialized first and its interrupt word will become non-NULL, while the first domain in `all_domains` order remains NULL, breaking part (2) of the invariant.

It is not completely clear to me what the right fix is, and the guidance of @gadmm would be much appreciated.

1. Maybe we don't actually need part (2) of the invariant, but it was added at the time because it held and made things easier to reason about. It may be that the property we need is that there is no *active/participating* domain after a NULL-interrupt-word domain in `all_domains` order; and I believe that this weaker property still holds. We could check it (cheaply) by (a) checking that all active/participating domains have a non-NULL interrupt word, and (b) calling `find_active_domain` on all non-NULL interrupt word following the first NULL interrupt word to ensure that they are not participating.

2. Maybe there *is* a problem if a non-NULL interrupt word is skipped by `caml_interrupt_all_signal_safe`, even if the corresponding domain is not currently active/participating. I'm not sure of what would be a good way to restore part (2) of the invariant if we do two-step reservation of the domain structure; it may be easier to adapt the runtime to work correctly if it stops being respected, but for this I would need to better understand what are the requirements around interrupt words.

gasche commented:
@gadmm #12410 is on my review pile, but I was hoping to get around to it only later (the fact that the change is subtle enough that you kept encoded exceptions is scary). The current PR stems from my obviously-faulty assumption that #14153 would be an easy to fix, and thus a good hacking distraction after the POPL deadline and before my holidays.

gadmm commented:
> (the fact that the change is subtle enough that you kept encoded exceptions is scary)

(no need to be scared, it is just that your new exception type has not been tested enough to be made convenient, and it sent me into a rabbit hole of actually making them nicer)

gadmm commented:
`caml_interrupt_all_signal_safe` uses the early exit to make sure that its loop is bounded by the max number of active domains in the program, rather than `caml_params->max_domains` which is an abstract number that can be chosen big.

I agree that it only requires that non-NULL interrupts words after some NULL belong to non-participating domains. However it might be harder to reason about it and to write an assertion to ensure it.

Note that you only need that the interrupt word is non-NULL, not that the domain is fully initialised. So all you need to do in theory is allocate the Caml_state as part of your first step. (I would be wary about allocating the Caml_state in a different processor than the new domain in case there are some of the NUMA effects that multicore devs sometimes talk about, but if this actually matters then there is already an issue due to the recycling of Caml_state between domains.)

gasche commented:
Thanks for the expert opinion. I will try to capture this invariant in the assertion and report back. And yes, I will try to look #12410 shortly (rather than in September, as would be reasonable).

gasche committed:
[minor] domain.c: _noexc version of find_stw_domain

gasche committed:
domain.c: refine the check_stw_domains invariant

The current invariant does not hold anymore if we move the reservation
of the domain structure to the caller of [domain_spawn] rather than
its callee: this lets two [domain_call] reserve domain structures
concurrently, and it could be that the first one to initialize its
domain has the second structure in [all_domains] order, breaking the
invariant that there is never a non-NULL interrupt word after a NULL
interrupt word.

Instead we check a refined invariant: there are no
active/participating domains after the first NULL in [all_domains]
order. The check is written to remain relatively efficient (despite
the fact that [find_active_domain] is linear; we avoid a quadratic
blowup in the common case).

gasche commented:
I updated the code with a refined invariant, checking that only non-active domains have a non-NULL interrupt word after the first NULL in `all_domains` order. This seems to work fine (on my machine, all tests pass in debug mode). The changes is in a separate commit towards the beginning of the PR.

gasche commented:
We discussed this at the "runtime meeting" today. There was consensus in favor of merging this, and @stedolan proposed to do a second review before that.

gasche commented:
@stedolan gentle ping.

My memory of our in-person discussion is that you wondered because this is so much code, My answer would be the following:
- Changing the minor-heap reservation requires a STW section to stop all mutators.
- The child domain being spawned cannot run a STW section as its STW state is not initialized, so I decided that the parent would be do it.
- Before, the domain-slot reservation happened in a child-initialization section that owns all_domains_lock and blocks any STW or parallel spawning from happening. After, it is possible that two domains would try to spawn in parallel, and this changes the invariants of the `all_domains` data structure.

It would be possible to split the PR in two, one half that only does this, changing who reserves the domain slot and weakening the invariants appropriates, and another PR that only touches the minor-heap reservation logic. (The two halves are cleanly separated as distinct commits in the current PR, which can be reviewed commit-by-commit.) Let me know if you would prefer this.

gasche commented:
"changes the invariants": the previous invariants were a-bit-subtle and undocumented. The new invariants are a-bit-subtle and documented; so of course there is a bit more code. The subtle part is the relation between the order of domain structures in `all_domains` and in `stw_domains.domains`, which `caml_interrupt_all_signal_safe` implicitly relies on.
